### PR TITLE
Added Fips 186.4 compliance, an additional strong Lucas-Selfridge (for BPSW) and a Frobenius (Paul Underwood) test, both optional. With documentation.

### PR DESCRIPTION
**Pull request by czurnieden**
(I hate GIT \*grrrr\* ;-) )

Three new methods are introduced in `mp_prime_is_prime`:

 -  a strong Lucas-Selfridge test which builds Thomas R. Nicely's famous BPSW test together with two Miller-Rabin tests with the bases 2 (two) and 3 (three) `int mp_prime_strong_lucas_selfridge(const mp_int *a, int *result)` A port of Thomas R. Nicely's code, see file `bn_mp_prime_strong_lucas_selfridge.c` for details regarding license (spoiler: all public domain).
 - a Frobenius test, the variation described by Paul Underwood<sup>4</sup> `int mp_prime_frobenius_underwood(const mp_int *N, int *result)`
 - a variable number of Miller-Rabin tests with random bases, Fips 186.4 (p. 126ff) compliant if the (P)RNG for `mp_rand` is cryptographically secure.

The method to use a variable number of Miller-Rabin tests with consecutive primes as bases has been shown to be insecure <sup>1,2,3</sup> and has been moved into a darker corner. It is still available with negative argument $t$ which offers the deterministic M-R test for $a<0x2be6951adc5b22410a5fd$ and runs the M-R test with at least the first 12 consecutive primes and up to PRIME\_SIZE if $(-t) > 13$ but finding that out needs a look into the documentation and that's something rarely done.

A positive argument $t$ will run $t$ random M-R tests and hence acts as the old version, just in a much safer way.

Both, the Frobenius and the Lucas-Selfridge tests need a Jacoby symbol that allows for negative input. To simplify things and to leave the API as it is the Kronecker symbol has been included by way of the function `mp_kronecker` to serve this need. Both tests need a function to pick and check a single bit, too. The function `mp_get_bit` does that (I wrote all others-set, clear, and toggle- already but did not want to bloat it). The last new function `mp_mul_si` does the same as `mp_mul_d` but allows for a signed small input (a `long`). It's single use is in `bn_mp_prime_strong_lucas_selfridge.c` and might be put there as a `static` function instead.

A larger problem: both the Frobenius and the Lucas-Selfridge tests do not work with `MP_8BIT` without some larger changes. But as a Fips 186.4 compliant method is included that works with `MP_8BIT` it should not matter much.

A smaller problem: both the Frobenius and the Lucas-Selfridge tests are quite slow in contrast to the Miller-Rabin test. They are currently placed between the first two M-R tests and the Fips/deterministic branches and although I don't think it is the best way to do it I don't know a better way to do so while keeping the API and accepting the fact that nobody reads the documentation.

The file `bn_mp_priome_is_prime.c` contains a couple of comments that go a little bit deeper into the specifics of the implementation.

Both the Frobenius and the Lucas-Selfridge tests are switched on for testing purposes (conditional preprocessor directives commented out).

Let me take the liberty to ask for an even greater scrutiny than already usual here because this is a change in a cryptographically sensitive part of the code. No, really, please check!

 
**1**: ARNAULT, François. Rabin-Miller primality test: composite numbers which pass it. Mathematics of Computation, 1995, 64. Jg., Nr. 209, S. 355-361.
**2**: ARNAULT, François. Constructing Carmichael numbers which are strong pseudoprimes to several bases. Journal of Symbolic Computation, 1995, 20. Jg., Nr. 2, S. 151-161.
**3**: BLEICHENBACHER, Daniel. Breaking a cryptographic protocol with pseudoprimes. In: International Workshop on Public Key Cryptography. Springer, Berlin, Heidelberg, 2005. S. 9-15.
**4**: UNDERWOOD, Paul. Quadratic Frobenius probable prime tests costing two selfridges. arXiv preprint arXiv:1706.01265, 2017.

### CODE REVIEW

**Comment by sjaeckel** [comment]
> (I hate GIT *grrrr* ;-) )

Git :P 

> Let me take the liberty to ask for an even greater scrutiny than already usual here because this is a change in a cryptographically sensitive part of the code. No, really, please check!

Good idea, I will! (but not now)

Probably @mkj and @karel-m want to have a look as well? Others are welcome as well!

**Comment by czurnieden** [comment]
The combined test (M-R with bases 2 and 3,  Frobenius (Underwood), and strong Lucas Selfridge) run against all base-2 Fermat pseudoprimes below  2^{64} passed, as expected. The ~2.2 GB large list is at http://www.cecm.sfu.ca/Pseudoprimes/psps-below-2-to-64.txt.bz2 and has 118,968,378 entries

The Fips test with 8 random bases (but with 64 bits, should be done with 8 bits because that's where it actually counts) had 2 failures (39672097809917653, 348621766474077253,), more or less as expected.

Code of the test program is at https://gist.github.com/czurnieden/4042ed0636e73b83453cc20c0f74ebf5

File: bn_mp_mul_si.c [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ * Michael Fromberger but has been written from scratch with
+ * additional optimizations in place.
+ *
+ * The library is free for all purposes without any express
+ * guarantee it works.
+ */
+
+// multiply bigint a with int d and put the result in c
+// Like mp_mul_d() but with a signed long as the small input
+int mp_mul_si(const mp_int *a, long d, mp_int *c)
```
[the remark this answers is not in the document]
Comment (czurnieden) [reply bn_mp_mul_si.c]: Your choice, just drop me a note and I'll change it. But I still think that function should better be put into the file `bn_mp_prime_frobenius_underwood.c` directly. Mmh...

File: bn_mp_kronecker.c [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+      if ((e = mp_div_2d(&a1, v, &a1, NULL)) != MP_OKAY) {
+         goto LBL_KRON;
+      }
+
+      if ((v & 0x1) == 1) {
+         k = k * table[p1.dp[0] & 7];
+      }
+
+      if (a1.sign == MP_NEG) {
+         // compute k = (-1)^((a1)*(p1-1)/4) * k
```
[the remark this answers is not in the document]
Comment (czurnieden) [reply bn_mp_kronecker.c]: Single line comments are in the C-standard since C-99 ;-)
But no problem.

**Comment by danaj** [comment]
I think this is a *huge* thing to have in libtommath.  A few years ago I wrote some libtommath code to do fast native-size tests (deterministic 32-bit Miller-Rabin using hashing, and BPSW for 64-bit) as well as BPSW for big numbers with an optional number of additional random base M-R tests for the extra paranoid.  It's been sitting on my drive.  :(

I have plans to spend more time on open source in July+ and this was near the top of my list.  Looks like I may be too late.  I'll take a look when I can.  The main problem I had was the Lucas test was surprisingly slow relative to a M-R test in libtommath, compared with GMP.

Thanks, @karel-m.

**Comment by czurnieden** [comment]
So, @danaj , under the highly preposterous assumption that the code I wrote is flawless does that mean I got a thumbs-up? ;-)

And you are right, the Lucas test is very slow in relation to MR but on the other side: it does a lot more and safety should come first in cryptographic code, optimization comes later (Yes, I know about timing attacks) and to be honest: it's not *that* bad.

I also put an  Frobenius-Underwood test in it because it works with `MP_8BIT` out of the box, at least to some extent, but it has roughly the same, large run-time. The Lucas test is more or less a direct port from Tom Nicely's implementation (with the stronger test for jacobi(a,n) == -1 instead of <=0) and might have some opportunities for optimization.

The Frobenius-Underwood test is ported from the pseudocode of the paper and I hope it is at least correct (I checked a vast amount of random numbers and primes against  your implementation in Math-Prime-Util-GMP-0.50). I didn't make any optimizations beside the obvious and I was even doubtful about the branch for `a!=0` (both `mul_d()` and ``add() have shortcuts) although I did not run any benchmarks and saw you have it in your version, too.

And the more I write about it the more I think I should drop that whole idea and replace it with a branch of the Lucas-test adapted to `MP_8BIT`. But for now: a F-U test followed by a bunch of random M-R tests is still better than just a bunch of random M-R tests. No matter what Fips says.

Default algorithm for now is

  1. Trial division, check for perf. square and so on
  2. Two M-R tests with the bases 2 (two) and 3 (three) respectively.
  3. A strong Lucas Selfridge test (or Frobenius-Underwood for MP_8BIT)
  4. At least one M-R test with a random base.

Basically BPSW with an extra random M-R test. The API stays, so it is still possible for the caller to ask for extra M-R tests but those tests have random bases. Only the Sorenson-Webster sequence up to base 41 (first 13 primes) is implemented for the deterministic test, nothing complicated.

**Comment by danaj** [comment]
@czurnieden sorry I have not looked closely and am temporarily moving to another country tomorrow so won't have time for a bit.  What I meant was that with native or GMP code, the Lucas test is something like 2x the cost of a M-R test, sometimes a little less.  With libtommath and mp_int I was seeing 5x or more.  I did not dig into why, and it was still a huge win vs. M-R with the 100 first primes, which is what the application I was targeting uses.  (which also has known counterexamples)

I definitely agree that F-U plus random-base M-R tests is stronger than the latter alone.  I think the various BPSW papers displayed good reasons why (Lucas tests in that case).

I did trial division (amount depends on bit length), M-R base 2, extra-strong Lucas which includes a perfect square test, then 0-n random-base M-R tests.  The base-3 test seems to be a Mathematica thing, as nobody else seems to do that.  It may be useful once optimizing performance to put one or two of the random-base M-R tests before the Lucas test given the speed difference (i.e. most base 2 pseudoprimes will get rejected faster, but for primes we run all the tests regardless so the order doesn't matter).

I'll see if I can find time to look at it, but it won't be in the next week.

**Comment by sjaeckel** [comment]
can you please rebase on top of the current develop?

**Comment by czurnieden** [comment]
Just as I was `pick`'ing and `squash`'ing the whole mess it came to me, quite suddenly, that such kind of method might not be exactly what you meant me to do. Please elaborate a bit more before I mess it up completely. Thank you in advance for your patience with an old man who had to learn more kinds of version-control than it has been deemed sane by the UN ;-)

**Comment by sjaeckel** [comment]
oh no sorry!

no squashing or else please, simply rebase all the commits :)

**Comment by czurnieden** [comment]
Mmh...
(origin=my fork at github, upstream=original libtommath at github)

```
 $  git rebase origin/develop 
Current branch develop is up to date.
```
As expected.
```
 $  git rebase upstream/develop 
First, rewinding head to replay your work on top of it...
Applying: Added Fips 186.4 compliance, an additional strong...
...
$ git push -f origin develop
Username for 'https://github.com': czurnieden
...
```
Correct?

**Comment by sjaeckel** [comment]
> Correct?

:+1:

**Comment by sjaeckel** [comment]
sorry, I just force-pushed the branch after rebasing on top of current develop

and then this wtf happened...!? to me it looks like a bug in @github that this issue is now closed...

**Comment by czurnieden** [comment]
The commit eebbcf06441e0d2380525adea4186d39c0525e5b "add mp_complement, mp_tc_div_2d, mp_tc_and, mp_tc_or, mp_tc_xor" (Mon Sep 10 18:48:58 2018 +0200)  for `demo/demo.c` introduced a shadow variable (`m`). At least according to the Travis-logs. 

Can somebody (maybe @minad ?) take a look at it? It looks innocent and may just be a residue of Steffen's mishap or my repairing but: better save than sorry as the saying goes, the reason might lie deeper.

File: bn_mp_kronecker.c [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+{
+   mp_int a1, p1, r;
+
+   int e = MP_OKAY;
+   int v, k;
+
+   const int table[8] = {0, 1, 0, -1, 0, -1, 0, 1};
+
+   if (mp_iszero(p)) {
+      if (a->used == 1 && a->dp[0] == 1) {
```
[the remark this answers is not in the document]
Comment (czurnieden) [reply bn_mp_kronecker.c]: Re: helper-pl results: I repaired the white-space etc. but what about the static functions? Not more than one function per file or no `static` functions at all? Both can be replaced with a rather ugly and barely readable macro if necessary, but is it?

The free'ing in `mp_kronecker` was a mess, indeed, thank you for pointing that out!

Testing for failures (although mainly allocation but others are possible, too) in libtommath is quite tedious. It would be really nice to have something to automate it. But that is just a "nice-to-have" and goes to the wishlist for 2020 ;-)

Re: using a macro for the `issmalldigit` kind of test&hellip;I do not think it would be a good idea because the digit `d` needs to be smaller than or equal to `MP_MASK`. That needs to be checked and I would advise to use `mp_cmp_d()` in such cases. On the other hand: a macro like
```c
#define  mp_isunity(a) (((a)->used == 1) && ((a)->dp[0] == 1))
```
might be useful but I already hear the complaints that it returns true for `-1`, too \*sigh\*.

My use here is correct because `kronecker(1,0) = kronecker(-1,0) = 1` and `kronecker(x,0) = 0` for all other `x` and a function call can be expensive. It would be no problem to use `mp_cmp_d()` instead if you so wish.

It is all locally ready for a `push` to this fork with the exception of the `static` functions.

[the remark this answers is not in the document]
Comment (sjaeckel) [reply bn_mp_kronecker.c]: >  but what about the static functions?

ah, we decided at one point that static functions should be prefixed with `_`

> wishlist for 2020

I appreciate your over-optimism ;-)

> My use here is correct because...

fine by me then as it is!

[the remark this answers is not in the document]
Comment (czurnieden) [reply bn_mp_kronecker.c]: > ah, we decided at one point that static functions should be prefixed with `_`

I really should have looked into the source of `helper.pl`, yes ;-)
But no problem, done.

No I get `first_comment` and `last_comment`
```
[first_comment] tommath_class.h
[last_comment] tommath_class.h
[last_comment] bn_mp_get_bit.c
[last_comment] bn_mp_kronecker.c
[last_comment] bn_mp_prime_frobenius_underwood.c
[last_comment] bn_mp_prime_strong_lucas_selfridge.c
check-comments:  FAIL 6
```
I added them manually to ` tommath_class.h` for now, should be put in the script in the near future. The rest passes now.
Will `push` now, let's see what comes up this time ;-)
(Oh, and please check the free'ings  in `kronecker.c`, thanks)

File: bn_mp_kronecker.c [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+   if ((e = mp_init_copy(&p1, p)) != MP_OKAY) {
+      goto LBL_KRON_0;
+   }
+
+   v = mp_cnt_lsb(&p1);
+   if ((e = mp_div_2d(&p1, v, &p1, NULL)) != MP_OKAY) {
+      goto LBL_KRON_1;
+   }
+
+   if ((v & 0x1) == 0) {
```
[the remark this answers is not in the document]
Comment (czurnieden) [reply bn_mp_kronecker.c]: Yes, that could be done with a table but I think the way how it is done now is the simplest way to do it and also without using up too much of precious stack. The latter can be helped additionally by changing the table to `signed char` instead of the `signed int` and casting the result but I do not think it would help a lot, that table does already fit into the smallest caches.

File: bn_mp_prime_frobenius_underwood.c [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+{
+   mp_int T1z,T2z,Np1z,sz,tz;
+
+   int a, ap2, length, i, j, isset;
+   int e = MP_OKAY;
+
+   *result = MP_NO;
+
+   if ((e = mp_init_multi(&T1z,&T2z,&Np1z,&sz,&tz, NULL)) != MP_OKAY) {
+      goto LBL_FU_ERR;
```
[the remark this answers is not in the document]
Comment (czurnieden) [reply bn_mp_prime_frobenius_underwood.c]: Yes, that slipped through, thanks!

File: bn_mp_kronecker.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (sjaeckel) [reply bn_mp_kronecker.c]: ah, I somehow missed the important point that the `if` operates on `v` but the table lookup operates on `a->dp[0]`
